### PR TITLE
Add max_time_ms feature to queryset

### PR DESCRIPTION
**Pull request by alexanderson**

### CODE REVIEW

**Comment by iainb** [comment]
I'm assuming this is a backport of mongoengine functionality. Would you be able to update the commit message to reference the commit refs?

File: tests/queryset.py [hunk 1]
```diff
@@ -3051,6 +3053,28 @@ class Number(Document):
 
         Number.drop_collection()
 
+    def test_max_time_ms(self):
```
Comment (iainb) [inline tests/queryset.py]: Is this the same test as upstream or is it an additional test? Might be worth adding in a comment that explains on line 3070 that this assumes that the query takes longer than 1ms to run and may flake.

File: mongoengine/queryset.py [hunk 2]
```diff
@@ -1057,6 +1060,16 @@ def timeout(self, yes_timeout):
         self._no_cursor_timeout = not yes_timeout
         return self
 
+    def max_time_ms(self, ms):
+        """Wait `ms` milliseconds before killing the query on the server
+        :param ms: the number of milliseconds before killing the query on the server
+        """
+        queryset = self.clone()
```
Comment (iainb) [inline mongoengine/queryset.py]: In our version of mongo engine most of the query set functions follow a pattern of:

```
if self._cursor_obj is not None:
    # apply change to cursor

# set attribute on current queryset
return self
```

I'm wondering if clone is required or if the pattern came from upstream. See `hint` or `batch_size` or `limit` functions on the `QuerySet` class.

Comment (alexanderson) [reply mongoengine/queryset.py]: @iainb 

Upstream this is not the case. Upstream clones the queryset on every operations it seems - which I would expect when chaining.
`my_queryset_2 = my_queryset_1.max_time_ms(1000)` I would expect `my_queryset_2` to have the max time set, but `my_queryset_1` to *not* have it set. But by modifying and returning `self`, it is set on both.

I found by making the direct changes from upstream introduces a dependency on the order in which things are chained. e.g. calling `Model.objects.max_time_ms(1000).order_by('field')` (like I do in the test) using the upstream changes meant that the `max_time_ms` was not propagated down the chain.

The changes I made were the only was to ensure correct chaining behaviour and persistence of value through further chaining.

**Pull request closed without merging**